pkg/tagset/base_factory.go

```diff
@@ -0,0 +1,43 @@
+package tagset
+
+// baseFactory provides some utility functions that are useful in all factory
+// implementations.
+type baseFactory struct {
+	// builders is a cache of unused builder instances for reuse
+	builders      []*Builder
```

Comment:
Why not use a `sync.Pool` for this? If there is a good reason not to use it, I think we should mention it in a comment

Reply:
`sync.Pool` is basically this, but with synchronization around it.  We don't need that synchronization since this is a single-thread data structure.

Reply:
`sync.Pool`'s implementation seems to be lock-free at least in the fast path (see e.g. golang/go@d5fd2dd6a17a816b7dfd99d4df70a85f1bf0de31). I definitely don't understand all the details but it's not obvious to me that this approach is clearly better. 

Anyway, not something to solve here, but maybe it warrants a comment?

Reply:
atomics are also synchronization primitives, and have a cost in terms of processor synchronization.

But, I can add a comment.

Reply:
Ah, I am thinking only of overhead, forgetting that `sync.Pool` also plays nicely with the gc and will drop all items when there is gc pressure.  I *think* there will only ever be one or two builders on this list, so not a great source of memory use, but we can modify it later if that assumption proves untrue.